diff --git a/minetest.conf.example b/minetest.conf.example
index eb883fa..1ac7865 100644
--- a/minetest.conf.example
+++ b/minetest.conf.example
@@ -38,7 +38,7 @@
 #keymap_left = KEY_KEY_A
 #keymap_right = KEY_KEY_D
 #keymap_jump = KEY_SPACE
-#keymap_sneak = KEY_LSHIFT
+#keymap_sneak = KEY_SHIFT
 #keymap_inventory = KEY_KEY_I
 #    Go down ladder / go down in fly mode / go fast in fast mode
 #keymap_special1 = KEY_KEY_E
diff --git a/src/defaultsettings.cpp b/src/defaultsettings.cpp
index 6504c51..f2b1c85 100644
--- a/src/defaultsettings.cpp
+++ b/src/defaultsettings.cpp
@@ -37,7 +37,7 @@ void set_default_settings(Settings *settings)
 	settings->setDefault("keymap_left", "KEY_KEY_A");
 	settings->setDefault("keymap_right", "KEY_KEY_D");
 	settings->setDefault("keymap_jump", "KEY_SPACE");
-	settings->setDefault("keymap_sneak", "KEY_LSHIFT");
+	settings->setDefault("keymap_sneak", "KEY_SHIFT");
 	settings->setDefault("keymap_drop", "KEY_KEY_Q");
 	settings->setDefault("keymap_inventory", "KEY_KEY_I");
 	settings->setDefault("keymap_special1", "KEY_KEY_E");
