--- store.old.lua	2014-05-23 15:28:39.000000000 +0200
+++ store.lua	2014-05-23 15:29:11.000000000 +0200
@@ -28,7 +28,7 @@
 	modstore.mods_on_search_page = searchmods
 	modstore.modsperpage = modstore.mods_on_unsorted_page
 
-	modstore.basetexturedir = core.get_texturepath() .. DIR_DELIM .. "base" ..
+	modstore.basetexturedir = "/Applications/minetest.app/Contents/Resources/bin/share/textures" .. DIR_DELIM .. "base" ..
 						DIR_DELIM .. "pack" .. DIR_DELIM
 
 	modstore.lastmodtitle = ""
