diff --git a/cmake/Modules/FindGettextLib.cmake b/cmake/Modules/FindGettextLib.cmake
index bc7ea36..78bc94b 100644
--- a/cmake/Modules/FindGettextLib.cmake
+++ b/cmake/Modules/FindGettextLib.cmake
@@ -1,21 +1,31 @@
 # Package finder for gettext libs and include files
 
 SET(CUSTOM_GETTEXT_PATH "${PROJECT_SOURCE_DIR}/../../gettext"
-	CACHE FILEPATH "path to custom gettext")
+	CACHE FILEPATH "/usr/local/opt/gettext")
 
 # by default
 SET(GETTEXT_FOUND FALSE)
 
 FIND_PATH(GETTEXT_INCLUDE_DIR
 	NAMES libintl.h
-	PATHS "${CUSTOM_GETTEXT_PATH}/include"
+	PATHS "${CUSTOM_GETTEXT_PATH}/include" "/usr/local/opt/gettext/include"
 	DOC "gettext include directory")
 
 FIND_PROGRAM(GETTEXT_MSGFMT
 	NAMES msgfmt
-	PATHS "${CUSTOM_GETTEXT_PATH}/bin"
+	PATHS "${CUSTOM_GETTEXT_PATH}/bin" "/usr/local/opt/gettext/bin"
 	DOC "path to msgfmt")
 
+FIND_LIBRARY(GETTEXT_LIBRARY
+	NAMES libintl.a
+	PATHS "${CUSTOM_GETTEXT_PATH}/lib" "/usr/local/opt/gettext/lib"
+	DOC "gettext *intl*.lib")
+
+FIND_LIBRARY(ICONV_LIBRARY
+	NAMES libiconv.dylib
+	PATHS "${CUSTOM_GETTEXT_PATH}/lib" "/usr/lib"
+	DOC "iconv lib")
+
 # modern Linux, as well as Mac, seem to not need require special linking
 # they do not because gettext is part of glibc
 # TODO check the requirements on other BSDs and older Linux
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 0ede2d6..28ac5ca 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -522,6 +522,7 @@ if(BUILD_CLIENT)
 		${PNG_LIBRARIES}
 		${X11_LIBRARIES}
 		${GETTEXT_LIBRARY}
+		${ICONV_LIBRARY}
 		${SOUND_LIBRARIES}
 		${SQLITE3_LIBRARY}
 		${LUA_LIBRARY}
